File: flash_test.go
package mineru_test

import (
	"context"
	"errors"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	mineru "github.com/OpenDataLab/mineru-open-sdk"
)

const (
	flashTestPDFURL  = "https://cdn-mineru.openxlab.org.cn/demo/example.pdf"
	flashTestBaseURL = "https://staging.mineru.org.cn/api/v1/agent"
	flashTestTimeout = 5 * time.Minute
)

// ═══════════════════════════════════════════════════════════════════
//  Unit tests — no API calls
// ═══════════════════════════════════════════════════════════════════

func TestNewFlash_NoTokenRequired(t *testing.T) {
	c := mineru.NewFlash()
	if c == nil {
		t.Fatal("NewFlash returned nil")
	}
}

func TestNewFlash_ExtractReturnsError(t *testing.T) {
	c := mineru.NewFlash()
	_, err := c.Extract(context.Background(), "https://example.com/doc.pdf")
	if err == nil {
		t.Fatal("expected error when calling Extract on flash-only client")
	}
	var paramErr *mineru.ParamError
	if !errors.As(err, &paramErr) {
		t.Fatalf("expected ParamError, got %T: %v", err, err)
	}
}

func TestNewFlash_CrawlReturnsError(t *testing.T) {
	c := mineru.NewFlash()
	_, err := c.Crawl(context.Background(), "https://example.com")
	if err == nil {
		t.Fatal("expected error when calling Crawl on flash-only client")
	}
	var paramErr *mineru.ParamError
	if !errors.As(err, &paramErr) {
		t.Fatalf("expected ParamError, got %T: %v", err, err)
	}
}

func TestNewFlash_SubmitReturnsError(t *testing.T) {
	c := mineru.NewFlash()
	_, err := c.Submit(context.Background(), "https://example.com/doc.pdf")
	if err == nil {
		t.Fatal("expected error when calling Submit on flash-only client")
	}
	var paramErr *mineru.ParamError
	if !errors.As(err, &paramErr) {
		t.Fatalf("expected ParamError, got %T: %v", err, err)
	}
}

func TestNewFlash_GetTaskReturnsError(t *testing.T) {
	c := mineru.NewFlash()
	_, err := c.GetTask(context.Background(), "fake-id")
	if err == nil {
		t.Fatal("expected error when calling GetTask on flash-only client")
	}
	var paramErr *mineru.ParamError
	if !errors.As(err, &paramErr) {
		t.Fatalf("expected ParamError, got %T: %v", err, err)
	}
}

func TestNew_FlashExtractAlsoWorks(t *testing.T) {
	if os.Getenv("MINERU_TOKEN") == "" {
		t.Skip("MINERU_TOKEN not set")
	}
	c, err := mineru.New("")
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	// Just verify the client has flash capability — don't actually call the API.
	_ = c
}

// ═══════════════════════════════════════════════════════════════════
//  Integration tests — flash API (no token needed)
// ═══════════════════════════════════════════════════════════════════

func newFlashTestClient() *mineru.Client {
	return mineru.NewFlash(mineru.WithBaseURL(flashTestBaseURL))
}

func flashExtractOrSkip(t *testing.T, c *mineru.Client, source string, opts ...mineru.FlashExtractOption) *mineru.ExtractResult {
	t.Helper()
	ctx, cancel := context.WithTimeout(context.Background(), flashTestTimeout)
	defer cancel()
	result, err := c.FlashExtract(ctx, source, opts...)
	if err != nil {
		errMsg := err.Error()
		if strings.Contains(errMsg, "405") || strings.Contains(errMsg, "404") || strings.Contains(errMsg, "503") {
			t.Skip("flash API not available (not deployed yet?)")
		}
		t.Fatalf("FlashExtract: %v", err)
	}
	return result
}

func TestFlashExtract_URL(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping integration test in short mode")
	}

	c := newFlashTestClient()
	result := flashExtractOrSkip(t, c, flashTestPDFURL, mineru.WithFlashPages("1-3"))
	if result.State != "done" {
		t.Fatalf("expected state=done, got %s (error: %s)", result.State, result.Error)
	}
	if result.Markdown == "" {
		t.Fatal("markdown is empty")
	}
	if result.TaskID == "" {
		t.Fatal("task_id is empty")
	}
}

func TestFlashExtract_LocalFile(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping integration test in short mode")
	}

	tmpDir := t.TempDir()
	localPDF := filepath.Join(tmpDir, "test_hello.pdf")
	if err := os.WriteFile(localPDF, minimalPDF, 0o644); err != nil {
		t.Fatalf("write test pdf: %v", err)
	}

	c := newFlashTestClient()
	result := flashExtractOrSkip(t, c, localPDF)
	if result.State != "done" {
		t.Fatalf("expected state=done, got %s (error: %s)", result.State, result.Error)
	}
	if result.Markdown == "" {
		t.Fatal("markdown is empty")
	}
}

func TestFlashExtract_SaveMarkdown(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping integration test in short mode")
	}

	c := newFlashTestClient()
	result := flashExtractOrSkip(t, c, flashTestPDFURL, mineru.WithFlashPages("1-1"))

	out := filepath.Join(t.TempDir(), "output.md")
	if err := result.SaveMarkdown(out, false); err != nil {
		t.Fatalf("SaveMarkdown: %v", err)
	}
	info, err := os.Stat(out)
	if err != nil {
		t.Fatalf("stat: %v", err)
	}
	if info.Size() == 0 {
		t.Fatal("saved markdown file is empty")
	}
}

func TestFlashExtract_NoDocxAvailable(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping integration test in short mode")
	}

	c := newFlashTestClient()
	result := flashExtractOrSkip(t, c, flashTestPDFURL, mineru.WithFlashPages("1-1"))

	if err := result.SaveDocx(filepath.Join(t.TempDir(), "out.docx")); err == nil {
		t.Fatal("expected error when saving docx from flash result")
	}
	if err := result.SaveHTML(filepath.Join(t.TempDir(), "out.html")); err == nil {
		t.Fatal("expected error when saving html from flash result")
	}
	if err := result.SaveLaTeX(filepath.Join(t.TempDir(), "out.tex")); err == nil {
		t.Fatal("expected error when saving latex from flash result")
	}
}

func TestFlashExtract_WithLanguage(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping integration test in short mode")
	}

	c := newFlashTestClient()
	result := flashExtractOrSkip(t, c, flashTestPDFURL,
		mineru.WithFlashLanguage("en"),
		mineru.WithFlashPages("1-1"),
	)
	if result.State != "done" {
		t.Fatalf("expected state=done, got %s", result.State)
	}
	if result.Markdown == "" {
		t.Fatal("markdown is empty")
	}
}

File: sdk/typescript/tests/flash.test.ts
/**
 * Flash mode tests — unit tests (no API) + integration tests (real API).
 */

import { describe, it, expect } from "vitest";
import { writeFile, mkdtemp, stat } from "node:fs/promises";
import { join } from "node:path";
import { tmpdir } from "node:os";
import {
  MinerU,
  NoAuthClientError,
  saveMarkdown,
  saveDocx,
  saveHtml,
  saveLatex,
} from "../src/index.js";

const FLASH_TEST_PDF_URL =
  "https://cdn-mineru.openxlab.org.cn/demo/example.pdf";
const FLASH_TEST_TIMEOUT = 300;

// ═══════════════════════════════════════════════════════════════════
//  Unit tests — no API calls
// ═══════════════════════════════════════════════════════════════════

describe("flash-only client (unit)", () => {
  const origToken = process.env["MINERU_TOKEN"];

  async function withoutToken(fn: () => Promise<void> | void): Promise<void> {
    delete process.env["MINERU_TOKEN"];
    try {
      await fn();
    } finally {
      if (origToken != null) {
        process.env["MINERU_TOKEN"] = origToken;
      }
    }
  }

  it("creates client without token", () => {
    withoutToken(() => {
      const c = new MinerU();
      expect(c).toBeDefined();
    });
  });

  it("extract throws NoAuthClientError", async () => {
    await withoutToken(async () => {
      const c = new MinerU();
      await expect(c.extract("https://cdn-mineru.openxlab.org.cn/demo/example.pdf")).rejects.toThrow(
        NoAuthClientError,
      );
    });
  });

  it("crawl throws NoAuthClientError", async () => {
    await withoutToken(async () => {
      const c = new MinerU();
      await expect(c.crawl("https://example.com")).rejects.toThrow(
        NoAuthClientError,
      );
    });
  });

  it("submit throws NoAuthClientError", async () => {
    await withoutToken(async () => {
      const c = new MinerU();
      await expect(c.submit("https://cdn-mineru.openxlab.org.cn/demo/example.pdf")).rejects.toThrow(
        NoAuthClientError,
      );
    });
  });

  it("getTask throws NoAuthClientError", async () => {
    await withoutToken(async () => {
      const c = new MinerU();
      await expect(c.getTask("fake-id")).rejects.toThrow(NoAuthClientError);
    });
  });
});

// ═══════════════════════════════════════════════════════════════════
//  Integration tests — flash API (no token needed)
// ═══════════════════════════════════════════════════════════════════

describe("flash extract URL", () => {
  const client = new MinerU();

  it("returns done with markdown", async () => {
    const result = await client.flashExtract(FLASH_TEST_PDF_URL, {
      pageRange: "1-3",
      timeout: FLASH_TEST_TIMEOUT,
    });

    expect(result.state).toBe("done");
    expect(result.markdown).not.toBeNull();
    expect(result.markdown!.length).toBeGreaterThan(0);
    expect(result.taskId).toBeTruthy();
  });

  it("with language option", async () => {
    const result = await client.flashExtract(FLASH_TEST_PDF_URL, {
      language: "en",
      pageRange: "1-1",
      timeout: FLASH_TEST_TIMEOUT,
    });

    expect(result.state).toBe("done");
    expect(result.markdown).not.toBeNull();
  });
});

describe("flash extract local file", () => {
  const client = new MinerU();

  it("local PDF returns markdown", async () => {
    const pdfBytes = Buffer.from(
      "%PDF-1.0\n" +
        "1 0 obj\n<< /Type /Catalog /Pages 2 0 R >>\nendobj\n" +
        "2 0 obj\n<< /Type /Pages /Kids [3 0 R] /Count 1 >>\nendobj\n" +
        "3 0 obj\n<< /Type /Page /Parent 2 0 R /MediaBox [0 0 612 792] " +
        "/Contents 4 0 R /Resources << /Font << /F1 5 0 R >> >> >>\nendobj\n" +
        "4 0 obj\n<< /Length 44 >>\nstream\n" +
        "BT /F1 24 Tf 100 700 Td (Hello MinerU) Tj ET\n" +
        "endstream\nendobj\n" +
        "5 0 obj\n<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>\nendobj\n" +
        "xref\n0 6\n" +
        "0000000000 65535 f \n" +
        "0000000009 00000 n \n" +
        "0000000058 00000 n \n" +
        "0000000115 00000 n \n" +
        "0000000266 00000 n \n" +
        "0000000360 00000 n \n" +
        "trailer\n<< /Size 6 /Root 1 0 R >>\n" +
        "startxref\n435\n%%EOF\n",
    );

    const dir = await mkdtemp(join(tmpdir(), "mineru-test-"));
    const path = join(dir, "test_hello.pdf");
    await writeFile(path, pdfBytes);

    const result = await client.flashExtract(path, {
      timeout: FLASH_TEST_TIMEOUT,
    });
    expect(result.state).toBe("done");
    expect(result.markdown).not.toBeNull();
  });
});

describe("flash extract save", () => {
  const client = new MinerU();

  it("save markdown works", async () => {
    const result = await client.flashExtract(FLASH_TEST_PDF_URL, {
      pageRange: "1-1",
      timeout: FLASH_TEST_TIMEOUT,
    });

    const dir = await mkdtemp(join(tmpdir(), "mineru-test-"));
    const out = join(dir, "output.md");
    await saveMarkdown(result, out, false);
    const s = await stat(out);
    expect(s.size).toBeGreaterThan(0);
  });

  it("save docx/html/latex raises", async () => {
    const result = await client.flashExtract(FLASH_TEST_PDF_URL, {
      pageRange: "1-1",
      timeout: FLASH_TEST_TIMEOUT,
    });

    await expect(
      saveDocx(result, join(tmpdir(), "out.docx")),
    ).rejects.toThrow();
    await expect(
      saveHtml(result, join(tmpdir(), "out.html")),
    ).rejects.toThrow();
    await expect(
      saveLatex(result, join(tmpdir(), "out.tex")),
    ).rejects.toThrow();
  });
});
